File: ShoppingBasket.Domain/Entities/Product.cs
using ShoppingBasket.Domain.Enums;
using ShoppingBasket.Domain.ValueObjects;

namespace ShoppingBasket.Domain.Entities;

public class Product : Entity
{
    public string Name { get; private set; }
    public Price Price { get; private set; }
    public string ImagePath { get; private set; } 
    public ProductCategory Category { get; private set; }
    
    public virtual List<Discount> Discounts { get; private set; }
    
    private Product(Guid id) : base(id) { }
    protected Product(Guid id, string name, Price price, string imagePath, ProductCategory category) : base(id)
    {
        Name = name;
        Price = price;
        ImagePath = imagePath;
        Category = category;
    }
    
    public static Product Create(string name, Price price, string imagePath, ProductCategory category)
    {
        return new Product(Guid.NewGuid(), name, price, imagePath, category);
    }
}

File: FE/ShoppingBasket.Presentation/src/app/models/product.ts
import { Discount } from "./discount";

export class Product {
    id!: string;
    name!: string;
    price!: number;
    imagePath!: string;
    discount: Discount[];
  
    constructor(data?: Partial<Product>) {
      Object.assign(this, data);
      if (data?.discount) {
        this.discount = data.discount.map(d => new Discount(d));
      } else {
        this.discount = [];
      }
    }
  }